pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -99,195 +107,215 @@ func genObservationID() (uint32, error) {
+		exp.process.CloseConnToCollector()
+		exp.process = nil
+		return
 	}
+	klog.V(2).Infof("Successfully exported IPFIX flow records")
+}
 
-	exp.process = expProcess
+func (exp *flowExporter) initFlowExporter() error {
+	var err error
+	exp.process, err = ipfix.NewIPFIXExportingProcess(exp.exporterInput)
+	if err != nil {
+		return fmt.Errorf("error when starting exporter: %v", err)
+	}
 	if exp.v4Enabled {
-		templateID := expProcess.NewTemplateID()
+		templateID := exp.process.NewTemplateID()
 		exp.templateIDv4 = templateID
-		if err := exp.set.PrepareSet(ipfixentities.Template, exp.templateIDv4); err != nil {
-			return fmt.Errorf("error when preparing set: %v", err)
+		if err = exp.ipfixSet.PrepareSet(ipfixentities.Template, exp.templateIDv4); err != nil {
+			return err
 		}
-		sentBytes, err := exp.sendTemplateSet(exp.set, false)
-		exp.set.ResetSet()
+		sentBytes, err := exp.sendTemplateSet(exp.ipfixSet, false)
+		exp.ipfixSet.ResetSet()
 		if err != nil {
 			return err
 		}
+
 		klog.V(2).Infof("Initialized flow exporter for IPv4 flow records and sent %d bytes size of template record", sentBytes)
 	}
 	if exp.v6Enabled {
-		templateID := expProcess.NewTemplateID()
+		templateID := exp.process.NewTemplateID()
 		exp.templateIDv6 = templateID
-		if err := exp.set.PrepareSet(ipfixentities.Template, exp.templateIDv6); err != nil {
-			return fmt.Errorf("error when preparing set: %v", err)
+		if err = exp.ipfixSet.PrepareSet(ipfixentities.Template, exp.templateIDv6); err != nil {
+			return err
 		}
-		sentBytes, err := exp.sendTemplateSet(exp.set, true)
-		exp.set.ResetSet()
+		sentBytes, err := exp.sendTemplateSet(exp.ipfixSet, true)
+		exp.ipfixSet.ResetSet()
 		if err != nil {
 			return err
 		}
 		klog.V(2).Infof("Initialized flow exporter for IPv6 flow records and sent %d bytes size of template record", sentBytes)
 	}
-
 	return nil
 }
 
 func (exp *flowExporter) sendFlowRecords() error {
-	addAndSendFlowRecord := func(key flowexporter.ConnectionKey, record flowexporter.FlowRecord) error {
-		exp.set.ResetSet()
+	updateOrSendFlowRecord := func(key flowexporter.ConnectionKey, record flowexporter.FlowRecord) error {
+		// We do not check for any timeout as the connection is still idle since
+		// the idleFlowTimeout was triggered.
+		if !record.IsActive {
+			return nil
+		}
+		// Send a flow record if the conditions for either timeout
+		// (activeFlowTimeout or idleFlowTimeout) are met. A flow is considered
+		// to be idle if its packet counts haven't changed since the last export.
+		if time.Since(record.LastExportTime) >= exp.idleFlowTimeout {
+			if ((record.Conn.OriginalPackets <= record.PrevPackets) && (record.Conn.ReversePackets <= record.PrevReversePackets)) || !record.Conn.IsPresent {
+				// Idle flow timeout
+				record.IsActive = false
+				goto sendRecord
+			}
+		}
+		if time.Since(record.LastExportTime) >= exp.activeFlowTimeout {
+			// Active flow timeout
+			goto sendRecord
+		}
+		return nil
+	sendRecord:
```

Comment:
The indentation is surprising to me here. Why is the `sendRecord` label aligned with the `updateOrSendFlowRecord` function declaration?

Also do we need a `goto` at all? Personally I think a boolean variable `recordNeedsSending` would make the code easier to read.

Reply:
done